core/src/main/scala/cats/Eval.scala

```diff
@@ -263,26 +263,13 @@ object Eval extends EvalInstances {
    * of FlatMap nodes) or "value" (in the case of Now, Later, or
    * Always nodes).
    */
-  @tailrec private def advance[A](fa: Eval[A]): Eval[A] =
-    fa match {
+  @tailrec private def advance[A](fa: Defer[A]): Eval[A] =
+    fa.thunk() match {
       case call: Eval.Defer[A] =>
-        advance(call.thunk())
-      case compute: Eval.FlatMap[A] =>
-        new Eval.FlatMap[A] {
-          type Start = compute.Start
-          val start: () => Eval[Start] = () => compute.start()
-          val run: Start => Eval[A] = s => advance1(compute.run(s))
```

Comment:
@non can you see why we did this? Can you write a test that shows the need? Removing it is a bit more efficient since we don't add a level of stack to all FlatMaps we call, BUT note, we should only really need to call advance when we know it is a Defer (note I change the argument type).